### ISSUE TITLE
Azure Blob connection string containing key with + chars fails

### ISSUE THREAD

**Issue by Dehalion** [aloneguid/storage#173]
If the key in a connection string for an Azure Blob Storage account contains a + character, the method
`AzureBlobStorageWithSharedKey` fails with the message "Invalid length for a Base-64 char array or string."

My rough guess is that the + characters are replaced with spaces somewhere (maybe there is some URL decoding).

Workaround (until fixed): Regenerate the key until you get one without a plus.

**Comment by Dehalion** [comment]
As I expected the `+` is replaced by a space in `StorageConnectionString.Parse()` with the line
```csharp
_parts[key] = value.UrlDecode();
```
which leaves me uncertain if this is a bug or a feature.

On one hand, it is formatted like a URI, starting with the storage type as "protocol". So one could argue `UrlDecode`ing the arguments is a valid thing to do.

On the other hand it is really unexpected for me as a user that I cannot just copy&paste the storage key into the connection string but have to check for characters which need URL encoding, even if its just a connection string, not an actual URL.

**Comment by banarualexandru** [comment]
I'm facing the same issue.
The workaround is like playing roulette...

**Comment by aloneguid** [comment]
This is unfortunate, yes plus sign in URL has a special meaning - space. We think of all the parameters as URL-encoded, because in some cases like drive path there are special characters that cause issues with a parser. The safest way to pass parameters is to url-encode them first though.

**Comment by Dehalion** [comment]
Yes, this is a bit unfortunate. It is inconsistent with how DB connection strings work, for example.
But I understand, why it is necessary to url-encode the parameters.

It would be great though if this is mentioned in the documentation.
I would add this as a hint to every occurrence of a connection string in the documentation, but I know this seems a bit excessive.

**Comment by 628426** [comment]
> Workaround (until fixed): Regenerate the key until you get one without a `+`.

> The workaround is like playing roulette...

Maybe instead of "key roulette" you could (strictly a workaround) use %2B instead of +

**Comment by Dehalion** [comment]
Yeah, should have updated the workaround after finding out the reason it doesn't work with a +. Have done it now.

**Comment by YannickRondeau** [comment]
@aloneguid should the connection string be encoded within the StorageNet library, or encoding it before initializing the storage would be the right way to do it?
